File: Chord Analyser/Metal/MetalRenderer.m
//
//  Apple Code
//

@import simd;
@import MetalKit;

#import "MetalRenderer.h"

@implementation MetalRenderer
{
    id<MTLDevice> _device;
    id<MTLCommandQueue> _commandQueue;
}

- (nonnull instancetype)initWithMetalKitView:(nonnull MTKView *)mtkView
{
    self = [super init];
    if (self) {
        _device = mtkView.device;
        _commandQueue = [_device newCommandQueue];
    }

    return self;
}

// Called whenever the view needs to render a frame.
- (void)drawInMTKView:(nonnull MTKView *)view
{
    // The render pass descriptor references the texture into which Metal should draw
    MTLRenderPassDescriptor *renderPassDescriptor = view.currentRenderPassDescriptor;
    if (renderPassDescriptor == nil) { return; }

    id<MTLCommandBuffer> commandBuffer = [_commandQueue commandBuffer];

    // Create a render pass and immediately end encoding, causing the drawable to be cleared
    id<MTLRenderCommandEncoder> commandEncoder = [commandBuffer renderCommandEncoderWithDescriptor:renderPassDescriptor];
    [commandEncoder endEncoding];
    
    // Get the drawable that will be presented at the end of the frame
    id<MTLDrawable> drawable = view.currentDrawable;

    // Request that the drawable texture be presented by the windowing system once drawing is done
    [commandBuffer presentDrawable:drawable];
    [commandBuffer commit];
}


// Called whenever view changes orientation or is resized
- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size {}

@end

File: Chord Analyser/Metal/MetalRenderer.swift
//
//  MetalRenderer.swift
//  Chord Analyser
//
//  Created by Jeremy Lindsay on 19/2/2023.
//

import Foundation
import simd
import MetalKit

class MetalRenderer : NSObject, MTKViewDelegate {
    
    private let device: MTLDevice
    private let commandQueue: MTLCommandQueue
    
    init(mtkView: MTKView) {
        self.device = mtkView.device!
        self.commandQueue = device.makeCommandQueue()!
    }
    
    func draw(in mtkView: MTKView) {
        
        // The render pass descriptor references the texture into which Metal should draw
        guard let renderPassDescriptor: MTLRenderPassDescriptor = mtkView.currentRenderPassDescriptor else { return }
        
        let commandBuffer: MTLCommandBuffer = self.commandQueue.makeCommandBuffer()!
        
        // Create a render pass and immediately end encoding, causing the drawable to be cleared
        let commandEncoder: MTLRenderCommandEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: renderPassDescriptor)!
        commandEncoder.endEncoding()
    
        // Get the drawable that will be presented at the end of the frame
        let drawable: MTLDrawable = mtkView.currentDrawable!
        
        // Request that the drawable texture be presented by the windowing system once drawing is done
        commandBuffer.present(drawable)
        commandBuffer.commit()
        
    }
    
    func mtkView(_ mtkView: MTKView, drawableSizeWillChange size: CGSize) {}
    
}
